beacon_node/http_api/src/lib.rs

```diff
@@ -1933,6 +1933,77 @@ pub fn serve<T: BeaconChainTypes>(
             })
         });
 
+    /*
+     * builder/states
+     */
+
+    let builder_states_path = eth_v1
+        .and(warp::path("builder"))
+        .and(warp::path("states"))
+        .and(chain_filter.clone());
+
+    // GET builder/states/{state_id}/expected_withdrawals
+    let get_expected_withdrawals = builder_states_path
+        .clone()
+        .and(warp::path::param::<StateId>())
+        .and(warp::query::<api_types::ExpectedWithdrawalsQuery>())
+        .and(warp::path("expected_withdrawals"))
+        .and(warp::path::end())
+        .and(warp::header::optional::<api_types::Accept>("accept"))
+        .and_then(
+            |chain: Arc<BeaconChain<T>>,
+             state_id: StateId,
+             query: api_types::ExpectedWithdrawalsQuery,
+             accept_header: Option<api_types::Accept>| {
+                blocking_response_task(move || {
+                    let proposal_slot = query.proposal_slot.unwrap_or(Slot::new(1));
+                    let (state, execution_optimistic, finalized) = match state_id.state(&chain) {
+                        Ok(state) => state,
+                        Err(_) => {
+                            return Err(warp_utils::reject::custom_not_found(
+                                "State not found".to_string(),
+                            ))
+                        }
+                    };
+                    let forkchoice_update_parameters = chain
+                        .canonical_head
+                        .cached_head()
+                        .forkchoice_update_parameters();
+
+                    let withdrawals = match chain.get_expected_withdrawals(
+                        &forkchoice_update_parameters,
+                        state.slot() + proposal_slot,
+                    ) {
```

Comment:
We shouldn't use `chain.get_expected_withdrawals` here, because it requires the `forkchoice_update_parameters`, which currently are being read from the head state rather than the state that was requested.

Instead, we should use the state that was loaded, and call `get_expected_withdrawals(&state, &chain.spec)`. Before computing the withdrawals this state should be _advanced_ to the epoch of the proposal slot using some logic like this:

```rust
let proposal_epoch = proposal_slot.epoch(E::slots_per_epoch());
if proposal_epoch != state.current_epoch() {
    partial_state_advance(
        &mut state,
        None,
        proposal_slot,
        &self.spec,
    )?;
}
```